src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -429,45 +429,53 @@ public JSObject CreateCSOwnedProxy(nint jsHandle)
             }
         }
 
-        public static void ReleaseCSOwnedObject(JSObject proxy, bool skipJS)
+        public static void ReleaseCSOwnedObject(JSObject jso, bool skipJS)
         {
-            if (proxy.IsDisposed)
+            if (jso.IsDisposed)
             {
                 return;
             }
-            var ctx = proxy.ProxyContext;
-#if FEATURE_WASM_MANAGED_THREADS
-            if (!ctx.IsCurrentThread())
-            {
-                throw new InvalidOperationException($"ReleaseCSOwnedObject has to run on the thread with same affinity as the proxy. ManagedThreadId: {Environment.CurrentManagedThreadId} JSHandle: {proxy.JSHandle}");
-            }
-#endif
+            var ctx = jso.ProxyContext;
+
             lock (ctx)
             {
-                if (proxy.IsDisposed)
+                if (jso.IsDisposed || ctx._isDisposed)
                 {
                     return;
                 }
-                var jsHandle = proxy.JSHandle;
-                proxy._isDisposed = true;
-                proxy.JSHandle = IntPtr.Zero;
-                GC.SuppressFinalize(proxy);
+                var jsHandle = jso.JSHandle;
+                jso._isDisposed = true;
+                jso.JSHandle = IntPtr.Zero;
+                GC.SuppressFinalize(jso);
                 if (!ctx.ThreadCsOwnedObjects.Remove(jsHandle))
                 {
                     Environment.FailFast($"ReleaseCSOwnedObject expected to find registration for JSHandle: {jsHandle}, ManagedThreadId: {Environment.CurrentManagedThreadId}. {Environment.NewLine} {Environment.StackTrace}");
                 };
-                if (!skipJS)
-                {
-                    Interop.Runtime.ReleaseCSOwnedObject(jsHandle);
-                }
                 if (IsJSVHandle(jsHandle))
                 {
                     ctx.FreeJSVHandle(jsHandle);
                 }
+                if (!skipJS)
```

Comment:
Is changing the order of `FreeJSVHandle` and `ReleaseCSOwnedObject` ok?

Reply:
This is great question

Reply:
I put it back  into correct order, but it would not work with async message of `ReleaseCSOwnedObjectPost`
But I'm not sure I need to solve that use-case. So I added `FailFast` for now.